src/toast/ops/filterbin.py

```diff
@@ -0,0 +1,1098 @@
+    legendre,
+)
+from ..mpi import MPI
+from ..observation import default_names as obs_names
+from ..pixels import PixelData, PixelDistribution
+from ..pixels_io import read_healpix_fits, write_healpix_fits
+from ..timing import Timer, function_timer
+from ..traits import Bool, Float, Instance, Int, Unicode, trait_docs
+from ..utils import Logger
+from .arithmetic import Subtract
+from .copy import Copy
+from .delete import Delete
+from .mapmaker_solve import SolverLHS, SolverRHS, solve
+from .mapmaker_utils import CovarianceAndHits
+from .operator import Operator
+from .pipeline import Pipeline
+from .pointing import BuildPixelDistribution
+from .scan_map import ScanMap, ScanMask
+
+
+def combine_observation_matrix(rootname):
+    """Combine slices of the observation matrix into a single
+    scipy sparse matrix file
+
+    Args:
+        rootname (str) : rootname of the matrix slices.  Typically
+            `{filterbin.output_dir}/{filterbin_name}_obs_matrix`.
+    Returns:
+        filename_matrix (str) : Name of the composed matrix file,
+            `{rootname}.npz`.
+    """
+
+    log = Logger.get()
+    timer0 = Timer()
+    timer0.start()
+    timer = Timer()
+    timer.start()
+
+    datafiles = sorted(glob(f"{rootname}.*.*.*.data.npy"))
+
+    all_data = []
+    all_indices = []
+    all_indptr = [0]
+
+    current_row = 0
+    current_offset = 0
+    shape = None
+
+    log.info(f"Combining observation matrix from {len(datafiles)} input files ...")
+
+    for datafile in datafiles:
+        parts = datafile.split(".")
+        row_start = int(parts[-5])
+        row_stop = int(parts[-4])
+        nrow_tot = int(parts[-3])
+        if shape is None:
+            shape = (nrow_tot, nrow_tot)
+        elif shape[0] != nrow_tot:
+            raise RuntimeError("Mismatch in shape")
+        if current_row != row_start:
+            all_indptr.append(np.zeros(row_start - current_row) + current_offset)
+            current_row = row_start
+        log.info(f"Loading {datafile}")
+        data = np.load(datafile)
+        indices = np.load(datafile.replace(".data.", ".indices."))
+        indptr = np.load(datafile.replace(".data.", ".indptr."))
+        all_data.append(data)
+        all_indices.append(indices)
+        indptr += current_offset
+        all_indptr.append(indptr[1:])
+        current_row = row_stop
+        current_offset = indptr[-1]
+
+    log.info_rank(f"Inputs loaded in", timer=timer, comm=None)
+
+    if current_row != nrow_tot:
+        all_indptr.append(np.zeros(nrow_tot - current_row) + current_offset)
+
+    log.info("Constructing CSR matrix ...")
+
+    all_data = np.hstack(all_data)
+    all_indices = np.hstack(all_indices)
+    all_indptr = np.hstack(all_indptr)
+    obs_matrix = scipy.sparse.csr_matrix((all_data, all_indices, all_indptr), shape)
+
+    log.info_rank(f"Constructed in", timer=timer, comm=None)
+
+    log.info(f"Writing {rootname}.npz ...")
+    scipy.sparse.save_npz(rootname, obs_matrix)
```

Comment:
Do we expect these files to have a long lifetime?  If so, then eventually we will need to define a format that is language-independent.  Obviously not an immediate concern!

Reply:
The intermediate files only exist until serially compiled into a single scipy sparse matrix (which still uses numpy save sets under the hood).